### PR TITLE
[FR Build a limited compatible rule ndjson for older stacks

### PR DESCRIPTION
**Pull request by Mikaayenson**
## Issues

#2808 

## Summary

Adds a CLI command to build a new ndjson file with all compatible rules for the user-supplied stack version.

## Testing

1. Command should build a file with rules that have fields only compatible with the version supplied. E.g. `setup` should field should not be included in rules with a stack 8.2.0 since that field is 8.3.0+.
2. Should be able to import the ndjson file into the stack after.

<details><summary>Sample CLI Command</summary>
<p>

```bash
(detection_dev) ➜  detection-rules git:(main) ✗ python -m detection_rules build-limited-rules --stack-version 8.3 --output-file releases/8.9/extras/limited-rules.ndjson
Loaded config file: /Users/stryker/workspace/ElasticGitHub/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Success: Rules written to releases/8.9/extras/limited-rules.ndjson
``` 

</p>
</details>

### CODE REVIEW

**Comment by eric-forte-elastic** [comment]
:green_circle: Test looks good :+1: 

<details><summary>Details</summary>
<p>

```shell
detection-rules on  2808-docs-add-a-document-with-steps-to-update-or-port-current-rules-to-717-stacks [$?] is  v0.1.0 via  v3.8.10 (detection-rules-build) on  eric.forte 
❯ python -m detection_rules build-limited-rules --stack-version 8.3 --output-file releases/8.9/extras/limited-rules.ndjson

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threat_match
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: machine_learning
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: threshold
Skipping unsupported rule type: threshold
Skipping unsupported rule type: new_terms
Skipping unsupported rule type: new_terms
Success: Rules written to releases/8.9/extras/limited-rules.ndjson
```

</p>
</details>

File: detection_rules/main.py [hunk 1]
```diff
@@ -113,6 +115,54 @@ def name_to_filename(name):
...
+            return
+
+        # Remove unsupported fields from rule
+        rule_contents = rule.contents.to_api_format()
+        for field in incompatible_fields:
+            rule_contents.pop(field, None)
+
+        # Try to write the cleaned rule to the NDJSON output file
+        try:
+            json.dump(rule_contents, file)
```
Comment (eric-forte-elastic) [inline detection_rules/main.py:153]: Should this use `utils.py`'s ndjson class .dump() ?

**Review by eric-forte-elastic — APPROVED** [review]
Few NITS, otherwise LGTM :+1:

File: detection_rules/main.py [hunk 2]
```diff
@@ -11,23 +11,25 @@
...
-from .rule import TOMLRule, TOMLRuleContents
+from .rule import TOMLRule, TOMLRuleContents, QueryRuleData
 from .rule_formatter import toml_write
 from .rule_loader import RuleCollection
-from .schemas import all_versions, definitions
+from .schemas import all_versions, definitions, get_incompatible_fields
 from .utils import get_path, get_etc_path, clear_caches, load_dump, load_rule_contents
 
 RULES_DIR = get_path('rules')
+UNSUPPORTED_RULE_TYPES = {'new_terms', 'threat_match', 'threshold', 'machine_learning'}
```
[the remark this answers is not in the document]
Comment (terrancedejesus) [reply detection_rules/main.py:32]: Agreed, let's define what we can in `definitions.py`

File: detection_rules/main.py [hunk 3]
```diff
@@ -113,6 +115,54 @@ def name_to_filename(name):
...
+    # Load all rules
+    rules = RuleCollection.default()
+
+    # Define output path
+    output_path = Path(output_file)
+
+    # Function to process and write each rule to the output file
+    def process_and_write_rule(rule, incompatible_fields: List[str], file):
+        if rule.contents.type in UNSUPPORTED_RULE_TYPES:
+            click.secho(f'Skipping unsupported rule type: {rule.contents.get("type")}', fg='yellow')
```
Comment (terrancedejesus) [inline detection_rules/main.py:143]:
```suggestion
            click.secho(f'{rule.name} - Skipping supported rule type: {rule.contents.get("type")}', fg='yellow')
```

**Review by terrancedejesus — APPROVED** [review]
Agreed with Eric on moving unsupported rule types.
Added a suggestion to reduce generic error.

File: detection_rules/main.py (hunk 1 above)
Comment (Mikaayenson) [reply detection_rules/main.py:153]: Which function are you thinking? I think the dump will write one at a time, whereas here we're iterating over multiple files writing to the same ndjson.

**Pull request merged**